plugins/com.google.cloud.tools.eclipse.appengine.standard.java8/src/com/google/cloud/tools/eclipse/appengine/standard/java8/FacetChangeListener.java

```diff
@@ -65,10 +65,12 @@ public void handleEvent(IFacetedProjectEvent event) {
       logger.warning(project + ": cannot find appengine-web.xml");
         return;
       }
-    if (project.hasProjectFacet(JavaFacet.VERSION_1_8)) {
-      AppEngineDescriptorTransform.addJava8Runtime(descriptor);
-    } else {
-      AppEngineDescriptorTransform.removeJava8Runtime(descriptor);
+    if (event.getType() == Type.POST_VERSION_CHANGE) {
```

Comment:
> what if appengine-web.xml already has <runtime>java8</runtime> before they add the facet?

Shouldn't it just work? When whatever facet is added, it examines the current state and does the right thing. That is, it checks if the Java facet is installed and if it does, it either adds or removes the `<runtime>` element depending on if it is a Java 8 facet or not. And the XSLT stylesheet doesn't seem like it will duplicate `<runtime>`. (However, I haven't tested all of these.)